### ISSUE TITLE
[BUG] Found bug fix for "stopped sending" or "stuck sending" for large campaigns

### ISSUE THREAD

**Issue by SEBreniac** [gophish/gophish#436]
### TLDR;
There is a race condition between `PostCampaign` and `GetQueuedCampaigns`

### Bug Explanation

In the `PostCampaign` function under models/campaign.go (line 246), the campaign is created and saved with the status `CAMPAIGN_QUEUED` (line 303). The loop on line 313 is used to create the results model for all of the targets associated with the campaign.

Since the campaign is already committed to the database with the status `CAMPAIGN_QUEUED`, the `GetQueuedCampaigns` call from the 1-minute polling function ends up finding this campaign ready to be executed. When it picks up the campaign in worker.go `processCampaign`, it begins to iterate over the campaigns results (c.Result : line 115).

For very large target lists, the `processCampaign` function picks up the c.Results value for the campaign BEFORE the `PostCampaign` loop to populate the results has finished. This leads to the `processCampaign` only sending emails for "part" of the campaign but thinks nothing is wrong when it completes the sending loop.

### Suggestion:
The quickest fix that I see to this is to do the following:
1.  Add a new const in models.go : CAMPAIGN_STAGED
2. In `PostCampaign`, save the initial campaign under the new status of `CAMPAIGN_STAGED`
3. After the results loops in `PostCampaign` (line 313), change the status of the campaign to `CAMPAIGN_QUEUED` and save. (This ensures that the result models for each target have been fully created before the campaign is kicked off)

### Temporary workaround:
Since this is a race condition, it is possible to work around this issue by simply using the scheduling feature and scheduling the campaign a decent bit in the future (5-15 minutes). Depending on the size of your target list, it will require a larger buffer window of time to complete the results loop before the campaign is picked up.

**Comment by jordan-wright** [comment]
This is great! Thanks for the awesome analysis. I'll investigate and will fix within the next couple of days.

**Comment by jordan-wright** [comment]
@SEBreniac I've added a fix for this. If you could do me a favor and give it a shot when you get a chance, I would appreciate it!

This doesn't fix the long GUI hangs when dealing with large group imports. I'm working that out as a separate issue. This should just fix the issue that occurs when launching a campaign with large groups.

**Comment by SEBreniac** [comment]
@jordan-wright I'm not sure that I'll be able to test the solution anytime soon (we finished the campaign we were doing). I did look over the code though, and it looks like it will address the issue. I'll be rebuilding from source.
